[CBRD-25947] ORDER BY clause is removed when querying ROWNUM in subqueries within IN-LIST

Hamkua opened the pull request:
[<http://jira.cubrid.org/browse/CBRD-25947>
](http://jira.cubrid.org/browse/CBRD-25947)
### Purpose

orderby_num()은 정렬 순서에 따라 결과가 달라지므로 ORDER BY 절이 필요합니다.

그러나 WHERE 절에 부질의가 사용되는 경우, 현재 부질의 내부의 ORDER BY 절이 제거되어 의도한 것과 다른 결과가 나오는 문제가 있습니다.

orderby_num() requires ORDER BY clause as its results depend on sort order.

However, there is an issue where ORDER BY clauses within subqueries in WHERE conditions are being removed, causing unexpected results.

### Implementation
where 절에 부질의가 사용된 경우, 부질의 전체에 대하여 distinct하다고 간주합니다.

재작성 단계에서 distinct 한 쿼리는 order by 절이 제거되도록 되어 있음에도, inst_num()을 orderby_num()으로 변환시키는 것이 원인입니다.

dinstinct 제약이 있을 때. inst_num() 을 orderby_num() 으로 재작성하는 것을 제한하도록 수정합니다.

When a subquery is used in WHERE clause, the entire subquery is considered DISTINCT.

This issue stems from converting inst_num() to orderby_num() during rewriting, despite the fact that DISTINCT queries are designed to have their ORDER BY clauses removed.

Prevent rewriting inst_num() to orderby_num() when query has DISTINCT constraint

### Remarks

N/A

Hamkua committed:
[CBRD-25947] modify to count sort column sequence numbers in queries that reference rownum

Hamkua committed:
[CBRD-25947] prevent rewriting inst_num() to orderby_num() when DISTINCT constraint exists

Hamkua commented on src/parser/view_transform.c:
```diff
@@ -2292,40 +2295,47 @@ mq_update_order_by (PARSER_CONTEXT * parser, PT_NODE * statement, PT_NODE * quer
 
-  /* generate orderby_num(), inst_num() */
-  if (!(ord_num = parser_new_node (parser, PT_EXPR)) || !(ins_num = parser_new_node (parser, PT_EXPR)))
+  if (order_by != NULL && statement->info.query.all_distinct == PT_DISTINCT)
```
중복값이 존재하는 컬럼을 기준으로 정렬하는 쿼리이면서, 해당 쿼리가 PT_DISTINCT 이면 쿼리 결과가 달라질 수 있습니다.
where 절에 부질의가 사용되었고, 정렬 순서에 영향을 받는 쿼리인 경우에는 PT_DISTINCT 대신 PT_ALL 로 변경해야 합니다.

**(3/5 주간회의) orderby_num()을 inst_num()으로 재작성하도록 수정하는 방법:**

inst_num() 으로 재작성할 orderby_num()이 이전에 한 번 변경된 것이 아니라 처음부터 orderby_num()으로 사용된 것이라면, inst_num()으로 변경 시 쿼리 결과가 달라질 수 있어 수정 중입니다.

아래의 두 쿼리는 결과가 다르게 출력됩니다.

* `select /*+ recompile */ orderby_num() from t2 order by c1 desc for orderby_num() between 2 and 4;`

* `select /*+ recompile */ rownum from t2 order by c1 desc for orderby_num() between 2 and 4;`

두 쿼리의 실행 결과가 다를 뿐더러, 쿼리 전체에 distinct 하다는 제약이 추가되면 정렬 기준 컬럼 값이 중복되는 경우 쿼리 결과가 달라지는 문제가 발생합니다.

* `select /*+ recompile */ orderby_num(), c1 from t2 order by c1 desc for orderby_num() between 2 and 4;`
* `select /*+ recompile */ distinct orderby_num(), c1 from t2 order by c1 desc for orderby_num() between 2 and 4;`

추가:
```sql
create table t1(c1 int, c2 int);
insert into t1 values(1,1),(2,2),(3,3),(4,4);

create table t2(c1 int);
insert into t2 values(1),(1),(2),(2);

-- orderby_num(), orderby_for를 포함하는 쿼리:
select /*+ recompile */ * from t1 where c1 in
(select  orderby_num() from t2 order by c1 desc for orderby_num() between 2 and 4);
```

youngjinj replied:
`ORDER BY`가 있는 질의에서 `rownum`과 `orderby_num()`의 채번은 다릅니다.

그래서 아래 2개 질의의 결과도 다릅니다.
사용자가 IN 조건의 부질의에 `rownum` 대신에 `orderby_num()`을 사용했다 하더라도
다른 질의로 봐야 하기 때문에 같은 결과가 조회되어야 한다고 생각하지는 않습니다.

```sql
create table t2(c1 int);
insert into t2 values(1),(1),(2),(2);

select rownum from t2 order by c1 desc for orderby_num() between 2 and 4;

                rownum
======================
                     3
                     2
                     1

select orderby_num() from t2 order by c1 desc for orderby_num() between 2 and 4;

         orderby_num()
======================
                     2
                     3
                     4
```

IN 조건에 대해서 `DISTINCT`를 추가하는 것에는 문제가 있는 것 같습니다.

rownum에 대해서는 아래 질의의 결과가 달라지지 않지만

```sql
select rownum, c1 from t2 order by c1;
select distinct rownum, c1 from t2 order by c1;

                rownum           c1
===================================
                     1            1
                     2            1
                     3            2
                     4            2
```

orderby_num()에 대해서는 아래 질의의 결과가 달라지기 때문입니다.

```sql
select orderby_num(), c1 from t2 order by c1;

         orderby_num()           c1
===================================
                     1            1
                     2            1
                     3            2
                     4            2

select distinct orderby_num(), c1 from t2 order by c1;
         orderby_num()           c1
===================================
                     1            1
                     2            2
```

IN 조건의 부질의에  `DISTINCT`를 추가할 때
orderby_num()이 있으면 제한하거나 한 번 더 부질의로 감싸고 `DISTINCT`를 추가하는 것에 대해서
고려되어야 필요하다고 생각합니다.

```sql
select /*+ recompile */ *
from t1
where c1 in (select distinct t.* from (select orderby_num() from t2 order by c1 desc for orderby_num() between 2 and 4) t);

           c1           c2
==========================
            2            2
            3            3
            4            4
```

shparkcubrid replied:
'order_by'가 필요 없어서 제거할 때 orderby_num()을 inst_num()으로 변경하라는 말이었는데요. 사용자가 직접 작성했더라도 order_by를 제거할때 orderby_num()을 남겨둘 수는 없을 것입니다.

'order_by'를 제거하면 결과가 달라지는 케이스라면 그 케이스에 대해서 order_by를 제거하지 않도록 해야하겠습니다.

Hamkua committed:
[CBRD-25947] modify queries to exclude DISTINCT when select-list has orderby_num() or when using ROWNUM with ORDER BY

shparkcubrid commented on src/parser/type_checking.c:
```diff
@@ -5259,20 +5259,45 @@ pt_coerce_range_expr_arguments (PARSER_CONTEXT * parser, PT_NODE * expr, PT_NODE
+      spec = arg2->info.query.q.select.from;
+      derived_spec = spec->info.spec.derived_table;
+
+      has_order_by =
```
회의때 이야기한 부분은 이 로직에서 order_by를 제거할 때 orderby_num()을 inst_num()으로 변경하는 이야기 였습니다.

Hamkua replied:
[f345cb7](https://github.com/CUBRID/cubrid/pull/5990/commits/f345cb7e216fa049277087fb7d1a71acdd273849) 에서 반영하였습니다.

Hamkua replied:
[33c73cc](https://github.com/CUBRID/cubrid/pull/5990/commits/33c73cc2c5e75e12da63f751739ec6f79a8bf17e) 

pt_coerce_range_expr_arguments 에서 pt_try_remove_order_by를 호출하지만,
order by는 pt_coerce_range_expr_arguments 가 호출되는 semantic check 단계가 아닌, 재작성 단계에서 제거될 수 있습니다.

pt_try_remove_order_by 함수는 order by가 제거된 이후에도 호출되므로, pt_try_remove_order_by 함수에서 orderby_num()을 inst_num() 으로 변경하도록 수정하였습니다.

shparkcubrid commented on src/parser/view_transform.c:
```diff
@@ -2152,10 +2152,13 @@ mq_update_order_by (PARSER_CONTEXT * parser, PT_NODE * statement, PT_NODE * quer
   assert (statement->node_type == PT_SELECT && query_spec->info.query.order_by != NULL);
 
   order_by = parser_copy_tree_list (parser, query_spec->info.query.order_by);
+  has_inst_or_orderby_num = pt_has_inst_or_orderby_num (parser, query_spec)
```
이부분은 다른 이슈에 대한 수정인것으로 보이는데 맞나요?
기존 문제 였던 in 연산자의 부질의일때 이 루틴을 사용하나요?

Hamkua replied:
where 절에 사용된 부질의가 distinct 하다는 점에 집중하다 보니, 해당 이슈의 범위를 벗어난 수정을 하게 된 것 같습니다.

distinct한 부질의가 인라인 뷰를 포함하고, 그 인라인 뷰가 order by, inst_num() 또는 orderby_num()을 포함하는 쿼리를 다루고자 수정하였으나, 불필요한 수정이므로 [f345cb7](https://github.com/CUBRID/cubrid/pull/5990/commits/f345cb7e216fa049277087fb7d1a71acdd273849) 에서 제거하였습니다.

``` sql
create table t1(c1 int, c2 int);
insert into t1 values(1,1),(2,2),(3,3),(4,4);

create table t2(c1 int);
insert into t2 values(1),(1),(2),(2);

select /*+ recompile */ * from t1 
    where c1 in (select n from (select rownum n from (select * from t2 order by 1)));

/* 
mq_update_order_by 직전 쿼리 구조
(select distinct [__t1].n from (select (orderby_num()), [__t1].c1 from [dba.t2] [__t1] order by 2) [__t1] (n))
*/
```

Hamkua committed:
[CBRD-25947] replace orderby_num() with inst_num() and remove ORDER BY clause in distinct queries

Hamkua committed:
[CBRD-25947] replace orderby_num() with inst_num() after removing ORDER BY in inline view

Hamkua committed:
[CBRD-25947] remove recursive calls and duplicate code

Hamkua committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-25947

Hamkua committed:
[CBRD-25947] prevent view merging when ORDER BY clause in inline view is expected to be eliminated for DISTINCT query

Hamkua committed:
[CBRD-25947] add specific conditions to prevent view merging

Hamkua committed:
Merge branch 'CUBRID:develop' into CBRD-25947

Hamkua committed:
[CBRD-25947] remove unnecessary comments

Hamkua commented on src/parser/view_transform.c:
```diff
@@ -1993,6 +2001,31 @@ mq_is_pushable_subquery (PARSER_CONTEXT * parser, PT_NODE * subquery, PT_NODE *
+   * In these cases, ORDER BY clause must be preserved by preventing view merging
+   * to maintain correct query results.
+   */
+  if (pt_is_distinct (mainquery) && subquery->info.query.order_by != NULL)
```
orderby_num() 이 아닌 일반 컬럼 조회 시에도 ORDER BY 절 제거로 인한 문제가 발생합니다.
``` sql
create table t1(c1 int, c2 int);
insert into t1 values(1,1),(2,2),(3,3),(4,4);

create table t2(c1 int);
insert into t2 values(1),(2),(3),(4);

-- 문제가 되는 쿼리
select /*+ recompile */ * from t2
    where c1 in (select c2 from (select c2 from t1 order by c1 desc) limit 2);

-- 예상 결과           실제 결과
--          c1                 c1
-- =============     =============
--           3                  1
--           4                  2
```
where 절에 사용된 쿼리는 다음과 같이 재작성됩니다.

``` sql
select distinct [__t1].c2 from t1 [__t1] for (orderby_num())<= ?:0 )
```
ORDER BY 제거 과정
`mq_update_order_by` 함수에서:
메인 쿼리가 distinct, 서브쿼리의 select-list에 정렬 기준 컬럼이 존재하지 않는 경우, 해당 정렬 조건은 메인 쿼리로 이동하지 않습니다.
뷰 머징 가능한 경우, 인라인 뷰의 ORDER BY는 제거되어야 합니다.

ORDER BY 절은 반드시 제거되어야 하는데, 그 이유는 다음과 같습니다.

정렬 기준 컬럼이 SELECT-LIST에 없으면 hidden-column으로 추가해야 합니다.
DISTINCT 제약이 있는 경우 SELECT-LIST 컬럼 추가 시 결과가 달라집니다.

이러한 이유로 ORDER BY 절 제거 방지 대신 뷰 머징을 제한하도록 수정하였으며, 다음 조건에서 뷰 머징 불가합니다:

* 메인쿼리의 SELECT-LIST에 ROWNUM/ORDERBY_NUM() 존재
* 메인쿼리에 LIMIT 절 사용
* 서브쿼리의 SELECT-LIST에 ORDERBY_NUM() 존재
* 서브쿼리에 LIMIT 절 사용

shparkcubrid replied:
```
/* subquery has order_by and main query has inst_num or analytic or order-sensitive aggrigation */
  if (subquery->info.query.order_by
      && ((!is_rownum_only && pt_has_inst_num (parser, pred)) || pt_has_analytic (parser, mainquery)
	  || pt_has_order_sensitive_agg (parser, mainquery) || pt_has_expr_of_inst_in_sel_list (parser, select_list)))
    {
      /* not pushable */
      return NON_PUSHABLE;
    }
```
1790라인 위 루틴과 중복같습니다. 메인질의에 distinct가 포함된경우를 제외하면 될 것같네요. is_rownum_only 는 오라클 스타일로 작성된 질의를 뷰머징 하기 위해 추가된 것입니다.
```
select * 
from (select ... order by 1,2) a
where rownum < 10;  <== is_rownum_only
```
그리고 조금 다른이야기로.. 부질의 order by를 무조건 보장해야 하는가에 대해서 생각해 봐야겠습니다. mq_update_order_by()의 로직은 부질의의 정렬을 최대한 보장하려고 하지만 불가한 경우 삭제하는 것으로 동작합니다. distinct, union 등의 연산은 정렬 혹은 해시 작업을 진행하기 때문에 부질의의 정렬이 유지되지 않을 수도 있습니다.

Hamkua committed:
[CBRD-25947] modify conversion logic between rownum and orderby_num based on ORDER BY existence

Hamkua committed:
[CBRD-25947] prevent view merging for inline views with distinct and orderby_num()

Hamkua committed:
[CBRD-25947] avoid converting LIMIT to ORDERBY_NUM() predicate in queries without ORDER BY

Hamkua committed:
[CBRD-25947] remove unnecessary INST_NUM to ORDERBY_NUM conversion for non-mergeable views

When view merging is prevented, there's no need to convert INST_NUM() to ORDERBY_NUM().
This commit removes this conversion and other related redundant code.

Hamkua committed:
[CBRD-25947] remove unnecessary code

Hamkua commented:
아래 나열된 SELECT 결과가 다르게 출력됩니다.

```sql
drop if exists t1, t2;

create table t1(c1 int, c2 int);
insert into t1 values(1,1),(2,2),(3,3),(4,4);

create table t2(c1 int);
insert into t2 values(1),(2),(3),(4);

select /*+ recompile */ c2 from t1
    where c1 in (select c2 from (select c2 from t1 order by c1 desc) limit 2);

/*
Query stmt:

select t1.c2 from t1 t1, (select distinct [__t1].c2 from (select t1.c2, t1.c1 from t1 t1 order by 2 desc ) [__t1] (c2, ha_2)
    where (inst_num()<= ?:0 )) av1861 (av_1) where t1.c1=av1861.av_1

           c2
=============
            3
            4

2 rows selected.
*/

select /*+ recompile */ c2 from t1
    where c1 in (select distinct c2 from (select c2 from t1 order by c1 desc) limit 2);

/*
Query stmt:

select t1.c2 from t1 t1, (select distinct [__t1].c2 from (select t1.c2, t1.c1 from t1 t1 order by 2 desc ) [__t1] (c2, ha_2) 
    for orderby_num()<= ?:0 ) av1861 (av_1) where t1.c1=av1861.av_1

           c2
=============
            1
            2

2 rows selected. 
*/

```

원인: 
where 절에 사용된 질의는 distinct가 추가되는데, 명시적으로 distinct를 추가한 쿼리는 limit 절을 inst_num() 으로 재작성,
그렇지 않은 쿼리는 limit 절을 orderby_num() 으로 재작성해 쿼리 결과가 다르게 출력됩니다.

order by 절이 제거되어 발생하는 문제가 아니며,
위와 같은 쿼리에서 부질의의 정렬 순서를 유지할 필요가 없다고 생각되어 현재 이슈에서 수정하지 않았습니다.

Hamkua committed:
[CBRD-25947] prevent merging when main query has DISTINCT and subquery uses rownum functions

Hamkua committed:
[CBRD-25947] modify to rewrite orderby_num() as inst_num() when ORDER BY changes in view merging cases
